internal/rpc/batcher.go

```diff
@@ -0,0 +1,86 @@
+package rpc
+
+import (
+	"context"
+	"math/big"
+	"sync"
+	"time"
+
+	gethRpc "github.com/ethereum/go-ethereum/rpc"
+	"github.com/rs/zerolog/log"
+	"github.com/thirdweb-dev/indexer/internal/common"
+)
+
+type RPCFetchBatchResult[T any] struct {
+	BlockNumber *big.Int
+	Error       error
+	Result      T
+}
+
+func RPCFetchInBatches[T any](rpc *Client, blockNumbers []*big.Int, batchSize int, batchDelay int, method string, argsFunc func(*big.Int) []interface{}) []RPCFetchBatchResult[T] {
+	if len(blockNumbers) <= batchSize {
+		return RPCFetchBatch[T](rpc, blockNumbers, method, argsFunc)
+	}
+	chunks := common.BigIntSliceToChunks(blockNumbers, batchSize)
+
+	log.Debug().Msgf("Fetching %s for %d blocks in %d chunks of max %d requests", method, len(blockNumbers), len(chunks), batchSize)
+
+	var wg sync.WaitGroup
+	resultsCh := make(chan []RPCFetchBatchResult[T], len(chunks))
+
+	for _, chunk := range chunks {
+		wg.Add(1)
+		go func(chunk []*big.Int) {
+			defer wg.Done()
+			resultsCh <- RPCFetchBatch[T](rpc, chunk, method, argsFunc)
+			if batchDelay > 0 {
+				time.Sleep(time.Duration(batchDelay) * time.Millisecond)
+			}
+		}(chunk)
+	}
+	go func() {
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	results := make([]RPCFetchBatchResult[T], 0, len(blockNumbers))
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
+	}
+
+	return results
+}
+
+func RPCFetchBatch[T any](rpc *Client, blockNumbers []*big.Int, method string, argsFunc func(*big.Int) []interface{}) []RPCFetchBatchResult[T] {
+	batch := make([]gethRpc.BatchElem, len(blockNumbers))
+	results := make([]RPCFetchBatchResult[T], len(blockNumbers))
+
+	for i, blockNum := range blockNumbers {
+		results[i] = RPCFetchBatchResult[T]{
+			BlockNumber: blockNum,
+		}
+		batch[i] = gethRpc.BatchElem{
```

Comment:
I excpect this to work with any node not only geth

Reply:
it's from the go ethereum library, but doesn't include anything specific to it. All JSON RPCs should accept the same format (it's usually the methods that differ)